tempesta_fw/msg.c

```diff
@@ -132,10 +132,18 @@ tfw_msg_iter_move(TfwMsgIter *it, unsigned char **data, unsigned long sz)
 
 		/* Linear skb part. */
 		if (it->frag < 0) {
+			unsigned char *begin = it->skb->data;
+			unsigned char *end = begin + skb_headlen(it->skb);
+			if (unlikely(addr < begin || addr > end))
+				addr = begin;
```

Comment:
Frankly, with the introduction of new variable `begin` the unchanged statement below now looks somewhat strange. It would be more logical if it looked like this now:
```
f_sz_rem = skb_headlen(it->skb) + begin - addr;
```

Same is true for a similar statement in the `else` branch.

Reply:
Thank you both, guys, this was dirty fixed. I have updated the `tfw_http_msg_insert()` function, so now it updates data pointer and following `tfw_msg_iter_move()` call gives correct iterator offset (not yet uploaded). But...

But I have spot a bug in `tfw_http_msg_insert()`, which is very simple wrapper around `ss_skb_get_room()`. The last function is called in many places. Look: `tfw_http_msg_insert()` inserts a 9-byte data piece at `TfwMsgIter {skb=ffff88977d9bd000, frag=8}`, at offset `ffff88976d004000`.

This the skb before insertion:
```
SKB (ffff88977d9bd000) DUMP: len=64858 data_len=64858 truesize=65882 users=1
  head=ffff88977d9bd100 data=ffff88977d9bd260 tail=160 end=1c0
  nr_frags=17 frag_list=          (null) next=ffff88977d241c00 prev=ffff88977d241000
  head data (0):
  frag 0 (addr=ffff8897713b92af pg_off=687 size=3409 pg_ref=3):
  frag 1 (addr=ffff88976d002000 pg_off=0 size=4096 pg_ref=2):
  frag 2 (addr=ffff88976d003000 pg_off=0 size=4096 pg_ref=2):
  frag 3 (addr=ffff88971c0a5000 pg_off=0 size=9 pg_ref=1):
  frag 4 (addr=ffff88977726c000 pg_off=0 size=4096 pg_ref=2):
  frag 5 (addr=ffff88977726d000 pg_off=0 size=4096 pg_ref=2):
  frag 6 (addr=ffff88977726e000 pg_off=0 size=4096 pg_ref=2):
  frag 7 (addr=ffff88977726f000 pg_off=0 size=4096 pg_ref=2):
  frag 8 (addr=ffff88976d004000 pg_off=0 size=4096 pg_ref=2):
  frag 9 (addr=ffff88976d005000 pg_off=0 size=4096 pg_ref=2):
  frag 10 (addr=ffff88976d006000 pg_off=0 size=4096 pg_ref=2):
  frag 11 (addr=ffff88976d007000 pg_off=0 size=4096 pg_ref=2):
  frag 12 (addr=ffff88974df1c000 pg_off=0 size=4096 pg_ref=2):
  frag 13 (addr=ffff88974df1d000 pg_off=0 size=4096 pg_ref=2):
  frag 14 (addr=ffff88974df1e000 pg_off=0 size=4096 pg_ref=2):
  frag 15 (addr=ffff88974df1f000 pg_off=0 size=4096 pg_ref=2):
  frag 16 (addr=ffff889770340000 pg_off=0 size=4096 pg_ref=2):
```
And after insertion:
```
SKB (ffff88977d9bd000) DUMP: len=60771 data_len=60771 truesize=61795 users=1
  head=ffff88977d9bd100 data=ffff88977d9bd260 tail=160 end=1c0
  nr_frags=17 frag_list=          (null) next=ffff88977d241c00 prev=ffff88977d241000
  head data (0):
  frag 0 (addr=ffff8897713b92af pg_off=687 size=3409 pg_ref=3):
  frag 1 (addr=ffff88976d002000 pg_off=0 size=4096 pg_ref=2):
  frag 2 (addr=ffff88976d003000 pg_off=0 size=4096 pg_ref=2):
  frag 3 (addr=ffff88971c0a5009 pg_off=9 size=9 pg_ref=2):
  frag 4 (addr=ffff88971c0a5000 pg_off=0 size=9 pg_ref=2):
  frag 5 (addr=ffff88977726c000 pg_off=0 size=4096 pg_ref=2):
  frag 6 (addr=ffff88977726d000 pg_off=0 size=4096 pg_ref=2):
  frag 7 (addr=ffff88977726e000 pg_off=0 size=4096 pg_ref=2):
  frag 8 (addr=ffff88977726f000 pg_off=0 size=4096 pg_ref=2):
  frag 9 (addr=ffff88976d004000 pg_off=0 size=4096 pg_ref=2):
  frag 10 (addr=ffff88976d005000 pg_off=0 size=4096 pg_ref=2):
  frag 11 (addr=ffff88976d006000 pg_off=0 size=4096 pg_ref=2):
  frag 12 (addr=ffff88976d007000 pg_off=0 size=4096 pg_ref=2):
  frag 13 (addr=ffff88974df1c000 pg_off=0 size=4096 pg_ref=2):
  frag 14 (addr=ffff88974df1d000 pg_off=0 size=4096 pg_ref=2):
  frag 15 (addr=ffff88974df1e000 pg_off=0 size=4096 pg_ref=2):
  frag 16 (addr=ffff88974df1f000 pg_off=0 size=4096 pg_ref=2):
```

The last insertion was not done at expected fragment, it was inserted as  fragment 3, instead of 8. This leads to broken h2 framing. Still looking for exact conditions. Happens only sometimes and only for some files for me.

Reply:
The insertion happens in a wrong place because of optimisation in `skb_fragment()`. 

Let's again take a look into the faulty skb:
```
SKB (ffff88977d9bd000) DUMP: len=64858 data_len=64858 truesize=65882 users=1
  head=ffff88977d9bd100 data=ffff88977d9bd260 tail=160 end=1c0
  nr_frags=17 frag_list=          (null) next=ffff88977d241c00 prev=ffff88977d241000
  head data (0):
  frag 0 (addr=ffff8897713b92af pg_off=687 size=3409 pg_ref=3):
  frag 1 (addr=ffff88976d002000 pg_off=0 size=4096 pg_ref=2):
  frag 2 (addr=ffff88976d003000 pg_off=0 size=4096 pg_ref=2):
  frag 3 (addr=ffff88971c0a5000 pg_off=0 size=9 pg_ref=1):
  frag 4 (addr=ffff88977726c000 pg_off=0 size=4096 pg_ref=2):
  frag 5 (addr=ffff88977726d000 pg_off=0 size=4096 pg_ref=2):
  frag 6 (addr=ffff88977726e000 pg_off=0 size=4096 pg_ref=2):
  frag 7 (addr=ffff88977726f000 pg_off=0 size=4096 pg_ref=2):
  frag 8 (addr=ffff88976d004000 pg_off=0 size=4096 pg_ref=2):
  frag 9 (addr=ffff88976d005000 pg_off=0 size=4096 pg_ref=2):
  frag 10 (addr=ffff88976d006000 pg_off=0 size=4096 pg_ref=2):
  frag 11 (addr=ffff88976d007000 pg_off=0 size=4096 pg_ref=2):
  frag 12 (addr=ffff88974df1c000 pg_off=0 size=4096 pg_ref=2):
  frag 13 (addr=ffff88974df1d000 pg_off=0 size=4096 pg_ref=2):
  frag 14 (addr=ffff88974df1e000 pg_off=0 size=4096 pg_ref=2):
  frag 15 (addr=ffff88974df1f000 pg_off=0 size=4096 pg_ref=2):
  frag 16 (addr=ffff889770340000 pg_off=0 size=4096 pg_ref=2):
```

 In the example we needed to insert a new data piece at `TfwMsgIter {skb=ffff88977d9bd000, frag=8}`, at offset `ffff88976d004000`. The `skb_fragment()` function doesn't take the `frag` as a hint (it's mostly unknown for callers anyway), and tries  to find the exact fragment. The function is happy, if the fragment ends just before the requested pointer, and makes an insertion there. In this example an insertion is requested before the  `ffff88976d004000` address, fragment 8. But fragment 2 ends just before the the same address. That is why insertion happens as fragment at unexpected place.

Why we never seen the issue before? We was lucky enough not to get skbs with such memory map in those rare functional tests which involves modification of extremely huge number/size of headers. 

We have no code depending on that behaviour, so I've removed the optimisation. Checked many times, the data received by a client now always match the original data.